File: binary-search/binary-search-test.rb
require 'simplecov'
SimpleCov.start
SimpleCov.command_name 'Unit Tests'

require "./binary-search.rb"
# require "../test/test.rb"
# include Test
require 'basic_test'
include BasicTest

test "BinarySearch.search, with element present",
  expected = 0,
  actual   = BinarySearch.search([1, 2, 3, 4, 5, 6, 7, 8, 9], 1)

test "BinarySearch.search, with element present",
  expected = 3,
  actual   = BinarySearch.search([1, 2, 3, 4, 5, 6, 7, 8, 9], 4)

test "BinarySearch.search, with element present",
  expected = 8,
  actual   = BinarySearch.search([1, 2, 3, 4, 5, 6, 7, 8, 9], 9)

test "BinarySearch.search, with element not present",
  expected = nil,
  actual   = BinarySearch.search([1, 2, 3, 4, 5, 6, 7, 8, 9], 21)

test "BinarySearch.search_for, with element present",
  expected = 2,
  actual   = BinarySearch.search_for([2, 4, 10, 10, 10, 18], 10, :first)

test "BinarySearch.search_for, with element not present",
  expected = nil,
  actual   = BinarySearch.search_for([2, 4, 10, 10, 10, 18], 21, :first)

test "BinarySearch.search_for, with element present, short array",
  expected = 0,
  actual   = BinarySearch.search_for([10], 10, :first)

test "BinarySearch.search_for, with element not present, short array",
  expected = nil,
  actual   = BinarySearch.search_for([10], 21, :first)

test "BinarySearch.search_for, with element present",
  expected = 4,
  actual   = BinarySearch.search_for([2, 4, 10, 10, 10, 18], 10, :last)

test "BinarySearch.search_for, with element not present",
  expected = nil,
  actual   = BinarySearch.search_for([2, 4, 10, 10, 10, 18], 21, :last)

test "BinarySearch.search_for, with element present, short array",
  expected = 0,
  actual   = BinarySearch.search_for([10], 10, :last)

test "BinarySearch.search_for, with element not present, short array",
  expected = nil,
  actual   = BinarySearch.search_for([10], 21, :last)

test "BinarySearch.count, with element present",
  expected  = 5,
  actual    = BinarySearch.count([2, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 10, 18], 10)

test "BinarySearch.count, with element not present",
  expected  = 0,
  actual    = BinarySearch.count([2, 4, 5, 6, 7, 8, 9, 10, 10, 10, 18], 11)

test "BinarySearch.rotations",
  expected  = 5,
  actual    = BinarySearch.rotations([5, 6, 7, 8, 9, 1, 2, 3, 4])

test "BinarySearch.rotations",
  expected  = 5,
  actual    = BinarySearch.rotations([2, 3, 4, 5, 6, 1])

test "BinarySearch.rotations",
  expected  = 1,
  actual    = BinarySearch.rotations([9, 1, 2, 3, 4, 5, 6])

test "BinarySearch.rotations, with no rotations",
  expected  = 0,
  actual    = BinarySearch.rotations([1, 2, 3, 4, 5, 6, 7, 8, 9])

test "BinarySearch.rotations, with duplicates",
  expected  = 9,
  actual    = BinarySearch.rotations([5, 6, 7, 8, 9, 9, 9, 9, 9, 1, 2, 3, 4])

test "BinarySearch.rotations, with iteration",
  expected  = 9,
  actual    = BinarySearch.rotations([5, 6, 7, 8, 9, 9, 9, 9, 9, 1, 2, 3, 4], method: :iteration)

test "BinarySearch.circular_search, when element is present",
  expected  = 7,
  actual    = BinarySearch.circular_search([5, 6, 7, 8, 9, 9, 1, 2, 3, 4], 2)

test "BinarySearch.circular_search, when element is not present",
  expected  = nil,
  actual    = BinarySearch.circular_search([5, 6, 7, 8, 9, 9, 1, 2, 3, 4], 11)

File: binary-search/java/BinarySearchTest.java
public class BinarySearchTest
{
  public static void main( String[] args )
  {
    int[] array = { 1, 2, 3, 4, 5, 6, 7, 8, 9 };
    int[] arrayWithDuplicates = { 2, 4, 10, 10, 10, 18 };
    int[] shortArray = { 10 };
    int[] arrayWithRotations = { 5, 6, 7, 8, 9, 1, 2, 3, 4 };
    int[] rotatedArrayWithDuplicates = { 5, 6, 6, 7, 8, 9, 1, 1, 3, 3, 3, 3, 4 };

    test (
      "BinarySearch.search, with element present",
      0,
      BinarySearch.search( array, 1 )
    );

    test (
      "BinarySearch.search, with element present",
      3,
      BinarySearch.search( array, 4 )
    );

    test (
      "BinarySearch.search, with element present",
      6,
      BinarySearch.search( array, 7 )
    );

    test (
      "BinarySearch.search, with element present",
      8,
      BinarySearch.search( array, 9 )
    );

    test (
      "BinarySearch.search, with element not present",
      -1,
      BinarySearch.search( array, 10 )
    );

    test (
      "BinarySearch.search_for, first instance, with element present",
      2,
      BinarySearch.search_for( arrayWithDuplicates, 10 , BinarySearch.Instance.FIRST )
    );

    test (
      "BinarySearch.search_for, first instance with element not present",
      -1,
      BinarySearch.search_for( arrayWithDuplicates, 11 , BinarySearch.Instance.FIRST )
    );

    test (
      "BinarySearch.search_for, last instance, with element present",
      4,
      BinarySearch.search_for( arrayWithDuplicates, 10 , BinarySearch.Instance.LAST )
    );

    test (
      "BinarySearch.search_for, last instance with element not present",
      -1,
      BinarySearch.search_for( arrayWithDuplicates, 11 , BinarySearch.Instance.LAST )
    );

    test (
      "BinarySearch.search_for, first instance, short array, with element present",
      0,
      BinarySearch.search_for( shortArray, 10 , BinarySearch.Instance.FIRST )
    );

    test (
      "BinarySearch.search_for, last instance, short array, with element present",
      0,
      BinarySearch.search_for( shortArray, 10 , BinarySearch.Instance.LAST )
    );

    test (
      "BinarySearch.count, with element present",
      3,
      BinarySearch.count( arrayWithDuplicates, 10 )
    );

    test (
      "BinarySearch.count, with element not present",
      0,
      BinarySearch.count( arrayWithDuplicates, 11 )
    );

    test (
      "BinarySearch.rotations, with no rotations, with recursion",
      0,
      BinarySearch.rotations( array, BinarySearch.Method.RECURSION )
    );

    test (
      "BinarySearch.rotations, with no rotations, with iteration",
      0,
      BinarySearch.rotations( array, BinarySearch.Method.ITERATION )
    );

    test (
      "BinarySearch.rotations, with no method",
      5,
      BinarySearch.rotations( arrayWithRotations )
    );

    test (
      "BinarySearch.rotations, with iteration",
      5,
      BinarySearch.rotations( arrayWithRotations, BinarySearch.Method.ITERATION )
    );

    test (
      "BinarySearch.rotations, with recursion",
      5,
      BinarySearch.rotations( arrayWithRotations, BinarySearch.Method.RECURSION )
    );

    test (
      "BinarySearch.rotations, with duplicates, with recursion",
      6,
      BinarySearch.rotations( rotatedArrayWithDuplicates, BinarySearch.Method.RECURSION )
    );

    test (
      "BinarySearch.rotations, with duplicates, with iteration",
      6,
      BinarySearch.rotations( rotatedArrayWithDuplicates, BinarySearch.Method.ITERATION )
    );

    test (
      "BinarySearch.circular_search, when element is present",
      7,
      BinarySearch.circular_search( arrayWithRotations, 3 )
    );

    test (
      "BinarySearch.circular_search, when element is not present",
      7,
      BinarySearch.circular_search( arrayWithRotations, 3 )
    );
  }

  static final String ANSI_RED = "\u001B[31m";
  static final String ANSI_RESET = "\u001B[0m";

  static void test( String testName, int expected, int actual )
  {
    double startTime = System.nanoTime();
    boolean result = expected == actual;
    double endTime = System.nanoTime();
    double durationInMicroseconds = (endTime - startTime) / 1000;

    if (result) {
      System.out.println(testName);
      System.out.println("PASS: (" + durationInMicroseconds + " μs)");
    } else {
      System.out.println(ANSI_RED + testName + ANSI_RESET);
      System.out.println(ANSI_RED + "Expected: " + expected + ANSI_RESET);
      System.out.println(ANSI_RED + "Actual:   " + actual + ANSI_RESET);
      System.out.println(ANSI_RED + "FAIL: (" + durationInMicroseconds + " μs)"  + ANSI_RESET);
    }
    System.out.println("----------------------------------");
  }
}
